arch: arm64: Use voting lock for multi-core boot race condition

povergoing opened the pull request:
Fixes the issue of multi-core booting mentioned in #63343 and probably fixes https://github.com/zephyrproject-rtos/zephyr/issues/57585#issuecomment-1540148084

The exclusive load/store instructions don't work well when MMU and cache are disabled on some cores e.g. Cortex-A7X. Change it to voting lock[1] to select the primary core when multi-cores boot simultaneously.

The voting lock has reasonable but minimal requirements on the memory system.

[1] https://www.kernel.org/doc/html/next/arch/arm/vlocks.html
Thanks @npitre for the suggestion

povergoing committed:
arch: arm64: Use voting lock for multi-core boot race condition

The exclusive load/store instructions don't work well when MMU and cache
are disabled on some cores e.g. Cortex-A72. Change it to voting lock[1]
to select the primary core when multi-cores boot simultaneously.

The voting lock has reasonable but minimal requirements on the memory
system.

[1] https://www.kernel.org/doc/html/next/arch/arm/vlocks.html

Signed-off-by: Jaxson Han <jaxson.han@arm.com>

npitre commented:
Actually you made a mistake. The following fix is needed:

```
diff --git a/arch/arm64/core/reset.S b/arch/arm64/core/reset.S
index 58dd41b78e..82d7cecd52 100644
--- a/arch/arm64/core/reset.S
+++ b/arch/arm64/core/reset.S
@@ -136,7 +136,7 @@ resetwait:
 	 */
 	get_cpu_logic_id x1, x2, x3, x4 //x1: MPID, x2: logic id
 
-	ldr	x4, [x0, #BOOT_PARAM_VOTING_OFFSET]
+	add	x4, x0, #BOOT_PARAM_VOTING_OFFSET
 
 	/* signal our desire to vote */
 	mov	w5, #1
```

And the current style in the file is to insert a tab between opcodes and their
first argument not a space.

Also, please add this in smp.c along the others:

```
BUILD_ASSERT(offsetof(struct boot_params, voting) == BOOT_PARAM_VOTING_OFFSET);
```

This to prevent the offset define from getting out of sync with the struct.

cocoeoli commented:
> Actually you made a mistake. The following fix is needed:
> 
> ```
> diff --git a/arch/arm64/core/reset.S b/arch/arm64/core/reset.S
> index 58dd41b78e..82d7cecd52 100644
> --- a/arch/arm64/core/reset.S
> +++ b/arch/arm64/core/reset.S
> @@ -136,7 +136,7 @@ resetwait:
>  	 */
>  	get_cpu_logic_id x1, x2, x3, x4 //x1: MPID, x2: logic id
>  
> -	ldr	x4, [x0, #BOOT_PARAM_VOTING_OFFSET]
> +	add	x4, x0, #BOOT_PARAM_VOTING_OFFSET
>  
>  	/* signal our desire to vote */
>  	mov	w5, #1
> ```
> 
> And the current style in the file is to insert a tab between opcodes and their first argument not a space.
> 
> Also, please add this in smp.c along the others:
> 
> ```
> BUILD_ASSERT(offsetof(struct boot_params, voting) == BOOT_PARAM_VOTING_OFFSET);
> ```
> 
> This to prevent the offset define from getting out of sync with the struct.

Yes, what should be load here is the address of the variable, not the value itself. The system can work well when MMU and cache are disabled after appling this code.

povergoing commented:
> Actually you made a mistake. The following fix is needed:
> 
> ```
> diff --git a/arch/arm64/core/reset.S b/arch/arm64/core/reset.S
> index 58dd41b78e..82d7cecd52 100644
> --- a/arch/arm64/core/reset.S
> +++ b/arch/arm64/core/reset.S
> @@ -136,7 +136,7 @@ resetwait:
>  	 */
>  	get_cpu_logic_id x1, x2, x3, x4 //x1: MPID, x2: logic id
>  
> -	ldr	x4, [x0, #BOOT_PARAM_VOTING_OFFSET]
> +	add	x4, x0, #BOOT_PARAM_VOTING_OFFSET
>  
>  	/* signal our desire to vote */
>  	mov	w5, #1
> ```

oops! Fixed it, I should have been more careful with assembly.

> 
> And the current style in the file is to insert a tab between opcodes and their first argument not a space.

Fixed with tab

> 
> Also, please add this in smp.c along the others:
> 
> ```
> BUILD_ASSERT(offsetof(struct boot_params, voting) == BOOT_PARAM_VOTING_OFFSET);
> ```
> 
> This to prevent the offset define from getting out of sync with the struct.

Good point! Fixed it.

povergoing commented:
> Yes, what should be load here is the address of the variable, not the value itself. The system can work well when MMU and cache are disabled after appling this code.

Really appreciate it

fabiobaltieri commented:
@povergoing should this be considered a blocking for 3.5?

cocoeoli commented:
```
	stlxr	w3, x1, [x0]
	/* If succeed, go to primary_core */
	cbz	w3, primary_core
	get_cpu_logic_id x1, x2, x3, x4 //x1: MPID, x2: logic id
```
Maybe use `/* */`  than  `//` .

povergoing commented:
@fabiobaltieri @jhedberg I don't think it is a blocker for 3.5.
